internal/eventstream/eventstream.go

```diff
@@ -0,0 +1,70 @@
+package eventstream
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/alecthomas/types/pubsub"
+)
+
+// EventStream is a stream of events that can be published and subscribed to, that update a materialized view
+type EventStream[T View] interface {
+	Publish(event Event[T]) error
+
+	View() T
+
+	Subscribe() <-chan Event[T]
+}
+
+// StreamView is a view of an event stream that can be subscribed to, without modifying the stream.
+type StreamView[T View] interface {
+	View() T
+
+	Subscribe() <-chan Event[T]
+}
+
+// View is a read-only view of the materialised current state of the event stream.
+type View interface {
+}
+
+type Event[T View] interface {
+
+	// Handle applies the event to the view
+	Handle(view T) (T, error)
```

Comment:
I mostly like the interface, except it seems weird for `Handle()` to mutate view _and_ return it. I think it would be less confusing if we picked one of:

1. `Handle(state T) error` - and `state` is mutated in place, which seems to be actually what's happening.
2. `Handle(view T) (T, error)` - and this returns a clone of `T`

1 seems preferable given the implementation.

Reply:
I will change to 2), but use reflect.DeepCopy to pass in a clone of T.